### PR TITLE
fix(ivy): enable packages/core/test/render3 test for AoT

### PR DESCRIPTION
**Pull request by mhevery**

### CODE REVIEW

**Review by IgorMinar — CHANGES_REQUESTED** [review]
I love having this flag to disable individual tests or blocks of test but I think it should be implemented differently.

Can you define the const in https://github.com/angular/angular/tree/master/packages/private/testing as a test helper because the const/flag is not a property of the compiler but rather a build configuration flag for our repo. You can still use the genrule to generate a file with the const, and reexport that const from https://github.com/angular/angular/blob/master/packages/private/testing/index.ts

You can then import the const from @angular/private/testing into the jasmine and karma init scripts and use the const in the body of the fixmeIvy function defined in these init scripts.

Also, as we discussed we no longer need to compile the framework in the JIT mode, so we can get rid of this test and simplify the signature of `ivyFixMe` to just `function ivyFixMe(reason: string): boolean`.

And lastly, can you rename this function from `ivyFixMe` to `fixmeIvy` or rename all bazel tags from `fixme-ivy-*` to `ivy-*-fixme`, so that we don't have two ways of calling this at different levels?

These changes simplify the setup and ivy, remove the need to pollute the production compiler code and the reliance on the flag published on the global ng object.

File: package.json [hunk 1]
```diff
@@ -144,4 +144,4 @@
     "vrsource-tslint-rules": "5.1.1",
     "webpack": "1.12.9"
   }
-}
+}
```
[the remark this answers is not in the document]
Comment (mhevery) [reply package.json]: reverted

File: packages/core/test/render3/pipe_spec.ts [hunk 2]
```diff
@@ -17,6 +17,14 @@ import {pipe, pipeBind1, pipeBind3, pipeBind4, pipeBindV} from '../../src/render
 import {RenderLog, getRendererFactory2, patchLoggingRenderer2} from './imported_renderer2';
 import {ComponentFixture, createComponent, getDirectiveOnNode, renderToHtml} from './render_util';
 
+const Directive: typeof _Directive = function(...args: any[]): any {
+  // In test we use @Directive for documentation only.
+  return () => undefined;
+} as any;
+const Pipe: typeof _Pipe = function(...args: any[]): any {
+  // In test we use @Component for documentation only.
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/test/render3/pipe_spec.ts]: fixed

File: tools/testing/init_node_spec.ts [hunk 3]
```diff
@@ -35,6 +35,7 @@ import 'zone.js/dist/jasmine-patch.js';
 
 (global as any).isNode = true;
 (global as any).isBrowser = false;
+(global as any).ivyFixMe = (mode: string) => mode !== (global as any).ng.ɵcompileMode;
```
[the remark this answers is not in the document]
Comment (mhevery) [reply tools/testing/init_node_spec.ts]: you don't need reason here as it is for reader only.

File: packages/compiler/BUILD.bazel [hunk 4]
```diff
@@ -29,3 +29,22 @@ ng_package(
         "//packages/compiler/testing",
     ],
 )
+
+## Controls if Ivy is enabled. (Temporary target until we permanently switch over to Ivy)
+##
+## This file generates `src/ivy_switch.ts` file which reexports symbols for `ViewEngine` or `Ivy.`
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/compiler/BUILD.bazel]: updated.

File: tools/testing/init_browser_spec.ts [hunk 5]
```diff
@@ -16,4 +16,5 @@ TestBed.initTestEnvironment(
 
 (window as any).isNode = false;
 (window as any).isBrowser = true;
-(window as any).global = window;
\ No newline at end of file
+(window as any).global = window;
+(window as any).ivyFixMe = () => true;
```
[the remark this answers is not in the document]
Comment (mhevery) [reply tools/testing/init_browser_spec.ts]: fixed.

File: packages/core/test/render3/pipe_spec.ts [hunk 6]
```diff
@@ -17,6 +17,14 @@ import {pipe, pipeBind1, pipeBind3, pipeBind4, pipeBindV} from '../../src/render
 import {RenderLog, getRendererFactory2, patchLoggingRenderer2} from './imported_renderer2';
 import {ComponentFixture, createComponent, getDirectiveOnNode, renderToHtml} from './render_util';
 
+const Directive: typeof _Directive = function(...args: any[]): any {
+  // In test we use @Directive for documentation only.
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/test/render3/pipe_spec.ts]: fixed

File: packages/core/test/render3/view_container_ref_spec.ts [hunk 7]
```diff
@@ -23,6 +23,12 @@ import {NgForOf} from '../../test/render3/common_with_def';
 import {getRendererFactory2} from './imported_renderer2';
 import {ComponentFixture, TemplateFixture, createComponent, getDirectiveOnNode} from './render_util';
 
+const Component: typeof _Component = function(...args: any[]): any {
+  // In test we use @Component for documentation only.
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/test/render3/view_container_ref_spec.ts]: fixed

File: packages/core/test/render3/load_domino.ts [hunk 8]
```diff
@@ -30,4 +30,5 @@ if (typeof window == 'undefined') {
   (global as any).Element = domino.impl.Element;
   (global as any).isBrowser = false;
   (global as any).isNode = true;
+  (global as any).ivyFixMe = (mode: string) => mode !== (global as any).ng.ɵcompileMode;
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/test/render3/load_domino.ts]: simplified.

File: packages/compiler/BUILD.bazel [hunk 9]
```diff
@@ -29,3 +29,22 @@ ng_package(
...
+## - append `--define=compile=legacy` (default) to `bazel` command to reexport `./ivy_switch_legacy`
+##   and use `ViewEngine`
+## - append `--define=compile=jit` to `bazel` command to rexport `./ivy_switch_jit` and use `Ivy`
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/compiler/BUILD.bazel]: dropped

File: packages/private/testing/src/fixme.ts [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * A global method which is used to conditionally block the execution of tests.
+ *
+ * ```
+ * fixmeIvy('some reason') && describe(...);
+ * ```
+ *
+ * The above will prevent the execution of the test(s) in Ivy mode, until they can be fixed.
+ */
+export function fixmeIvy(reason?: string): boolean {
```
Comment (IgorMinar) [inline packages/private/testing/src/fixme.ts]: can we make the reason required?

Comment (IgorMinar) [reply packages/private/testing/src/fixme.ts]: even "unknown reason" is better than no reason at all.

**Review by IgorMinar — APPROVED** [review]
Thanks for making all the changes. I left one more suggestion but the rest looks great!

File: .circleci/config.yml [hunk 11]
```diff
@@ -122,21 +122,6 @@ jobs:
...
-          <<: *post_checkout
-      - restore_cache:
-          key: *cache_key
-      - *define_env_vars
-      - *setup_circleci_bazel_config
-      - *yarn_install
-      - *setup_bazel_remote_execution
-
-      - run: yarn test-ivy-jit //...
-
```
Comment (IgorMinar) [inline .circleci/config.yml]: oops.. there are more references to "test_ivy_jit" in this file that need to be removed.

**Comment by IgorMinar** [comment]
and more more thing: to make cherry-picking easier we are trying to merge all major infrastructure changes to master and patch branch, so I relabeled this PR to master&patch.

**Comment by kara** [comment]
@mhevery @IgorMinar Looks like the test_ivy_jit task is still failing with `Cannot find a job named `test_ivy_jit` to run in the `jobs:` section of your configuration file.`

**Comment by kara** [comment]
FYI we will be creating a new PR for patch

**Comment by mhevery** [comment]
@IgorMinar I tried to patch this to `7.0.x` but the two branches have diverged too much to be able to do that. The issue is that `master` no longer has `ivy_switch` and the `7.0.x` still does.

**Comment by IgorMinar** [comment]
OK. If it's too much work then it's ok to skip, but in that case can you at
least cherry pick the circle ci changes to reduce the load ok the ci?
Thanks

**Pull request closed without merging**